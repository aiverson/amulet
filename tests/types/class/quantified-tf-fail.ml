class show 'a begin
  val show : 'a -> string
end
external val (^) : string -> string -> string = ""

class app 'f begin
  type apply 'a
end

type proxy 'a = Proxy

instance show (proxy 'a) begin
  let show _ = "Proxy"
end

instance app proxy begin
  type apply 'a = proxy 'a
end

type ex_ty_fun 'f =
  | Mk : forall 'f 'a. app 'f => apply 'f 'a -> ex_ty_fun 'f

instance (forall 'a. show (apply 'f 'a)) => show (ex_ty_fun 'f) begin
  let show (Mk x) = "Mk " ^ show x
end
