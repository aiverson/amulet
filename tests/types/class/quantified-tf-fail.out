quantified-tf-fail.ml[24:29 ..24:34]: error (E2018)
  Could not deduce show (apply 'f 'a) from the context
    forall 'a. show (apply 'f 'a)

  • Note: this constraint can not be quantified over
    because it originated from a method in an instance.

  • Possible fix: add it to the instance context

  Arising in the the expression
   │ 
24 │   let show (Mk x) = "Mk " ^ show x
   │                             ^^^^^^
  • Note: apply is an associated type, and so may not be injective
